### PR TITLE
Add energy balance equation to Compositional solver.

### PR DESCRIPTION
**Pull request by CusiniM**
This original content of the PR has been split into multiple pieces that have been implemented separately:
- PR 1: Add a model for solid internal energy  (https://github.com/GEOSX/GEOSX/pull/1660)
- PR 2: Add the enthalpy and internal energy functions to the CO2-brine fluid model (https://github.com/GEOSX/GEOSX/pull/1689)
- PRs 3, 4, 5: Refactor accumulation, flux, and property kernels (https://github.com/GEOSX/GEOSX/pull/1696, https://github.com/GEOSX/GEOSX/pull/1672, and https://github.com/GEOSX/GEOSX/pull/1698)
- PR 6: Add thermal conductivity constitutive model (https://github.com/GEOSX/GEOSX/pull/1732)
- PR 7: Reduce the size of the fluid kernel wrapper (https://github.com/GEOSX/GEOSX/pull/1779)
- PR 8: Put everything together in a `CO2BrineThermal` fluid model (https://github.com/GEOSX/GEOSX/pull/1755)
- This PR: Add the energy balance equation to the compositional solver

This PR implements the kernel interfaces that add the thermal derivatives and assemble the energy balance (internal energy term, enthalpy flux, conduction flux). 

- [X] Unit test (finite difference derivatives)
- [X] Integrated test (CO2-brine simulation)

Rebaseline in https://github.com/GEOSX/integratedTests/pull/213

### CODE REVIEW

**Comment by CusiniM** [comment]
Conduction is still missing. I see two way to add it:

1. add another stencil with it's own coefficient which would be the conductivity.
2. allow the stencil to have more than one coefficient. Basically I would modify the stencil so that it could contain weights relative to two different properties (or coefficients). We may want to modify some of the naming if we choose this path.

I personally prefer option 2 because the connections for the advective and the conductive parts are exactly the same so we would only have to loop over the connections once instead of twice. 

@francoishamon @rrsettgast @klevzoff @castelletto1 opinions?

File: src/coreComponents/constitutive/fluid/MultiFluidBase.cpp
Comment (francoishamon) [inline src/coreComponents/constitutive/fluid/MultiFluidBase.cpp:81]: @klevzoff @rrsettgast 
In this PR, I added some vectors in `MultiFluidBase.hpp`  (8 `arraysViews` for phase enthalpy and phase internal energy) and I ended up with a [problem that we had some time ago](https://github.com/GEOSX/GEOSX/pull/1325#discussion_r577888918) with the CO2-Brine fluid model: I cannot compile using CUDA for some compilers (it is fine on Lassen for some reason) because of a "formal parameter space overflow" in the fluid update. 

For example [here](https://app.travis-ci.com/github/GEOSX/GEOSX/jobs/554920358), I got:
```
/opt/GEOSX_TPL/raja/include/RAJA/policy/cuda/forall.hpp(137): Error: Formal parameter space overflowed (4392 bytes required, max 4096 bytes allowed) in function _ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256ENS_9Iterators16numeric_iteratorIllPlEEZN5geosx9PVTDriver7runTestINS8_12constitutive29MultiPhaseMultiComponentFluidINSB_8PVTProps19EzrokhiBrineDensityENSD_21EzrokhiBrineViscosityENSD_20SpanWagnerCO2DensityENSD_20FenghourCO2ViscosityENSD_13CO2SolubilityEEEEEvRT_RN7LvArray9ArrayViewIdLi2ELi1ElNSM_10ChaiBufferEEEEUliE_lEEvT1_T0_T2_
```
In the case above, the error takes place in a unit test, but I also get it when the fluid update kernel is called from `CompositionalMultiphaseBaseKernels.hpp` on Pecan.

For this combination of CO2-brine models, we have to pass to the fluid update GPU kernel:
- 32 `arrayViews`
- 7 `TableFunction::KernelWrappers` (ultimately 9 when we do the enthalpy and internal energy updates)

Is there an approach that would allow us to overcome this issue? 
If the `TableFunction::KernelWrappers` are too large, should we go back to the (not very beautiful) `arrayView1d< TableFunction::KernelWrappers >` of size 1?

I am tagging @andrea-borio who had the same issue for the `MultiphasePoromechanicsKernel` with VEM.

Comment (TotoGaz) [reply src/coreComponents/constitutive/fluid/MultiFluidBase.cpp:81]: Thx @klevzoff 
 
> 1. Combine derivative arrays for `dP`, `dT`, `dC` into one array, having the same number of dimensions as `dC`, but with +2 in runtime size. [...] The downsides are slight reduction in readability and a minor performance regression on CPU in non-thermal cases since temperature values will be force-read into cache but remain unused (GPU should not be affected because due to different memory layout temperature values are effectively separated from the rest).

Is it possible, instead of providing a table, to provide a `class` holding this "_augmented_" array and providing accessors to the `dP`, `dT`, `dC` data? Something _like_
```c++
DT = 0; DP = 1; ...
double SomeClass::dT( int a, int b, int c ) const {
    this->array[ DT ][ a ][ b ][ c ];
}
double SomeClass::dP( int a, int b, int c ) const {
    this->array[ DP ][ a ][ b ][ c ];
}
```
so we can do `fields->dT( a, b, c )`? Would it still help about the memory issues we are speaking about?

In terms of readability, I'm not sure it's a real issue. In terms of memory layout performance, I'm sure we can `template meta-programming` something 🤣 

> 2. Move `TableFunction` wrappers into global memory (via `array1d` storage or similar), saving ~640 bytes (80 bytes per table). The downside are all the additional dynamic allocations than will need to happen on every kernel launch [...]

Are we (you 🤣 ) able to estimate the time of these additional allocation?
Is it possible, if we go this way, to implement dedicated "container" (instead of the `array1d` trick) so the intent would be clearer?

> 3. Create some sort of class that aggregates together multiple array views of equal size, saving on not storing repeated dims and strides. Depending on how we group arrays, we could save in the range of 256-1024 bytes (hard to estimate precisely). I'm struggling a bit to imagine what the interface and usage of that class would look like.

Do you have in mind something like I write in `1.` or is it something completely different?
Actually, would an extended version of https://github.com/GEOSX/GEOSX/pull/1750 (not serving the arrays, but accessing the data) be an example? If it's the case, would keeping up in https://github.com/GEOSX/GEOSX/pull/1750 's direction help at encapsulating, before we can change the contracts?

Comment (francoishamon) [reply src/coreComponents/constitutive/fluid/MultiFluidBase.cpp:81]: Thanks very much for these suggestions @klevzoff and @TotoGaz, I am going to try Suggestion 1. (combining derivative arrays) in a separate PR so we can discuss something concrete soon.

Comment (klevzoff) [reply src/coreComponents/constitutive/fluid/MultiFluidBase.cpp:81]: > Do you have in mind something like I write in `1.` or is it something completely different?

What I had in mind there is somewhat similar, but implemented in `LvArray` and thus devoid of any physical interpretation, e.g. (simplifying a bit)
```c++
template< int NUM_ARRAYS, typename T, int NDIM, typename INDEX_TYPE, typename PERM, typename BUFFER >
class MultiArrayView
{
public:
  template< typename ... INDICES >
  T & operator( int const array_index, INDICES ... indices )
  {
     return m_buffers[ array_index ][ linearIndex( indices ... ) ];
  }
private:
  INDEX_TYPE m_dims[NDIM];    // we store this once instead if NUM_ARRAYS times
  INDEX_TYPE m_strides[NDIM]; // ditto
  BUFFER m_buffers[NUM_ARRAYS];
};
...
// packs together views into all derivative arrays of matching size
MultiArrayView< 3, real64, 4, ... > dPhaseProps_dComp = ...
real64 const dPhaseDens_dC = dPhaseProps_dComp( 0, k, q, ip, ic );
real64 const dPhaseVisc_dC = dPhaseProps_dComp( 1, k, q, ip, ic );
real64 const dPhaseMassDens_dC = dPhaseProps_dComp( 2, k, q, ip, ic );
```
Of course, we can then add a GEOSX-side wrapper that gives "physical" names to those 0, 1, 2 indices.

Comment (klevzoff) [reply src/coreComponents/constitutive/fluid/MultiFluidBase.cpp:81]: > Is it possible, instead of providing a table, to provide a `class` holding this "_augmented_" array and providing accessors to the `dP`, `dT`, `dC` data? Something _like_
> ...
> so we can do `fields->dT( a, b, c )`? Would it still help about the memory issues we are speaking about?

Yes, to be precise, it could look something like (here I also mashed values with derivatives, which may or may not be desired):
```c++
template< int NDIM, typename PERM >
class CompositionalVarView
{
public:
  template< typename ... INDICES >
  real64 & value( INDICES ... indices ) 
  { return m_view( indices..., 0 ); } // 0 is offset of value

  template< typename ... INDICES >
  real64 & dP( INDICES ... indices ) 
  { return m_view( indices..., 1 ); } // 1 is offset of pressure derivative

  template< typename ... INDICES >
  real64 & dT( INDICES ... indices ) 
  { return m_view( indices..., 2 ); } // 1 is offset of temperature derivative

  template< typename ... INDICES >
  real64 & dT( integer jc, INDICES ... indices ) // jc can't be after the pack, which is unfortunate
  { return m_view( indices..., 3 + jc ); } // 3 is offset of composition derivatives

private:
  // +1 accounts for extra derivative dimension
  ArrayView< real64, NDIM + 1, PERM > m_view;
}
...
CompositionalVarView< 3, multifluid::LAYOUT_PHASE > density = ...
forall( k, q, ip, jc )
{
  real64 const dens = density.value( k, q, ip );
  real64 const dDens_dP = density.dP( k, q, ip );
  real64 const dDens_dT = density.dT( k, q, ip );
  real64 const dDens_dC = density.dC( jc, k, q, ip ); // :sad_face:
}
```

Unlike the previous reply, this involves changing the actual storage of the data. But it results in more logical grouping IMO.

**Comment by francoishamon** [comment]
@CusiniM GitHub prevents me from selecting your name as a reviewer because you opened this PR, but it is now ready for review. I added an 2D integrated test with cold CO2 injection in brine, and I will now work on validation against another code.

I am also going to add a unit test to validate the derivatives, but I am going to refactor the unit tests `testCompMultiphaseFlow` and `testCompMultiphaseHybrid` first to avoid code duplication (in a separate PR).

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseUtilities.hpp
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseUtilities.hpp:46]: I am not sure I fully understand what's going on here.

File: src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp:506]: why twice? I m guessing it's just to pass smt to the function coz thta should be the derivative wrt to pressure.

File: src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp:517]: this is only the conductive part right?

**Review by CusiniM — COMMENTED** [review]
Outstanding job @francoishamon ! It's a pretty big PR:
- we should probably organize a walk through
- we should allow you to merge asap. maintaining it up-to-date with develop is going to be painful.

File: src/coreComponents/constitutive/fluid/multiFluidSelector.hpp
Comment (francoishamon) [reply src/coreComponents/constitutive/fluid/multiFluidSelector.hpp:48]: Yes, I really don't understand what is going on there:
- If I add only `CO2BrinePhillipsThermalFluid`, the code compiles and works
- If I add only `CO2BrineEzrokhiThermalFluid`, the code compiles and works
- If I add both, I get the following error:
```
[ 61%] Building CUDA object coreComponents/physicsSolvers/CMakeFiles/physicsSolvers.dir/fluidFlow/CompositionalMultiphaseBase.cpp.o                                                      
[ 63%] Building CUDA object coreComponents/physicsSolvers/CMakeFiles/physicsSolvers.dir/fluidFlow/wells/CompositionalMultiphaseWell.cpp.o                                                
[ 63%] Building CUDA object coreComponents/constitutive/CMakeFiles/constitutive.dir/fluid/PVTDriver.cpp.o                                                                                
[ 88%] Built target constitutive                                                                                                                                                         
nvcc error   : 'ptxas' died due to signal 11 (Invalid memory reference)                                                                                                                  
make[3]: *** [coreComponents/physicsSolvers/CMakeFiles/physicsSolvers.dir/build.make:115: coreComponents/physicsSolvers/CMakeFiles/physicsSolvers.dir/fluidFlow/CompositionalMultiphaseBase.cpp.o] Error 11 
```

This is not a blocking problem for me, but I don't understand why it does not compile with the two models activated.

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:1176]: Yes I agree. It it not yet done in the PR, but I will do it.

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseUtilities.hpp
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseUtilities.hpp:62]: Here I modified the code shifting the rows and computing the row sum of mass equations because when there is an energy equation, we don't want to include it in the sum, and we don't want to shift it.

In the latest commit, I tried to document what the function does with more doxygen that explains the thermal case:

The `localFluxJacobian` looks like that:
```
 *                       p    \rho_1   \rho_2   T
 *    (mass_comp_1)_1    .       .       .      .       \
 *    (mass_comp_2)_1    .       .       .      .       |=> block 1
 *    (energy)_1         .       .       .      .       /
 *
 *    (mass_comp_1)_2    .       .       .      .       \
 *    (mass_comp_2)_2    .       .       .      .       |=> block 2
 *    (energy)_2         .       .       .      .       /
 ```
and we have
```
 *      - numRowsToShift = 2
 *      - numRowsInBlock = 3
 *      - numColsInBlock = 4
 *      - numBlocks = 2
 ```
So, in each block, the code is going to sum the `mass_comp_X` equations, shift them, but will just skip the `energy` equation. This was the reason for adding the argument `numRowsToShift` to the function.

File: src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp]: For thermal conductivity, I implemented what the two of us discussed some time ago: the dependence of thermal conductivity on saturation and porosity is treated explicitly to make sure that we can easily reuse the `computeWeights` function. So yes you are right, I pass `m_thermalConductivity` twice because I need to pass something, but I don't have anything else to pass. I added more explanations in the code.

In a future PR, we could modify the `computeWeights` function to accommodate explicit coefficients  (just pass the coefficient, and return the transmissibility). That would be useful for thermal transmissibility and also for diffusion transmissibility that would likely be explicit/lagged in iteration as well.

File: src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp:522]: Yes you are right, this is only the conductive part here, added to the energy flux. The enthalpy flux transported by the phases is computed and added to the energy flux in the lambda above. I added some comments to clarify that.

File: src/coreComponents/constitutive/solid/CoupledSolidBase.hpp
Comment (francoishamon) [inline src/coreComponents/constitutive/solid/CoupledSolidBase.hpp:136]:
```suggestion
```
Not an intentional change, I have to remove this, I don't use it anymore

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:1166]: I addressed this comment [here](https://github.com/GEOSX/GEOSX/pull/1827).

**Review by klevzoff — APPROVED** [review]
Looks great!

File: src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp]: Yes, thanks for the suggestion, I changed this to `[&]` and passed the `stack` by reference, which solved my problems.

File: src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp]: Here as well, I replaced `[=]` with `[&]`.

File: inputFiles/thermal/co2_thermal_2d.xml
Comment (francoishamon) [reply inputFiles/thermal/co2_thermal_2d.xml]: Yes, I agree, I created a folder called `thermalMultiphaseFlow` in `inputFiles` and in the integrated tests repository.

File: src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseFVMKernels.hpp]: In the latest. commit, I put PascalCase here

File: src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseBaseKernels.hpp
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseBaseKernels.hpp]: Yes I agree. I removed the `REDUCE_POLICY` and the `LOCAL_VECTOR` templates in `ResidualNormKernel` and `SolutionCheckKernel` for single-phase/compositional flow and well kernels.

File: src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseBaseKernels.hpp
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/ThermalCompositionalMultiphaseBaseKernels.hpp:505]: I think it would be great to try, if everyone agrees. I can post an issue to get feedback and try after the PR is merged.

**Pull request merged**
